File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/diagnosiskeydist/SubmissionRepositoryTest.java
package uk.nhs.nhsx.diagnosiskeydist;

import com.amazonaws.services.s3.model.S3ObjectSummary;
import org.junit.jupiter.api.Test;
import uk.nhs.nhsx.testhelper.data.TestData;
import uk.nhs.nhsx.diagnosiskeydist.s3.SubmissionFromS3Repository;
import uk.nhs.nhsx.diagnosiskeyssubmission.model.StoredTemporaryExposureKey;
import uk.nhs.nhsx.diagnosiskeyssubmission.model.StoredTemporaryExposureKeyPayload;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;
import java.util.Date;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static uk.nhs.nhsx.testhelper.data.TestData.STORED_KEYS_PAYLOAD_DESERIALIZED;
import static uk.nhs.nhsx.testhelper.data.TestData.STORED_KEYS_PAYLOAD_DESERIALIZED_DAYS_SINCE_ONSET;

public class SubmissionRepositoryTest {

    @Test
    public void deserializesStoredPayload() throws IOException {
        InputStream inputStream = new ByteArrayInputStream(TestData.STORED_KEYS_PAYLOAD.getBytes());

        StoredTemporaryExposureKeyPayload payload = SubmissionRepository.getTemporaryExposureKeys(inputStream);
        assertThat(payload.temporaryExposureKeys).isNotEmpty();
        assertThat(payload.temporaryExposureKeys.stream().allMatch(it -> matchKey(it, STORED_KEYS_PAYLOAD_DESERIALIZED))).isTrue();
    }

    @Test
    public void deserializesStoredPayloadWithDaysSinceOnset() throws IOException {
        InputStream inputStream = new ByteArrayInputStream(TestData.STORED_KEYS_PAYLOAD_DAYS_SINCE_ONSET.getBytes());

        StoredTemporaryExposureKeyPayload payload = SubmissionRepository.getTemporaryExposureKeys(inputStream);
        assertThat(payload.temporaryExposureKeys).isNotEmpty();
        assertThat(payload.temporaryExposureKeys.stream().allMatch(it -> matchKey(it, STORED_KEYS_PAYLOAD_DESERIALIZED_DAYS_SINCE_ONSET))).isTrue();
    }

    private boolean matchKey(StoredTemporaryExposureKey storedKey, StoredTemporaryExposureKeyPayload deserializedPayload) {
        return deserializedPayload.temporaryExposureKeys.stream()
                .anyMatch(k ->
                        k.key.equals(storedKey.key) &&
                                k.rollingPeriod.equals(storedKey.rollingPeriod) &&
                                k.rollingStartNumber.equals(storedKey.rollingStartNumber) &&
                                k.transmissionRisk.equals(storedKey.transmissionRisk) &&
                                ((k.daysSinceOnsetOfSymptoms == null && storedKey.daysSinceOnsetOfSymptoms == null) ||
                                        (k.daysSinceOnsetOfSymptoms != null && k.daysSinceOnsetOfSymptoms.equals(storedKey.daysSinceOnsetOfSymptoms)))
                );
    }

    private long now = System.currentTimeMillis();

    @Test
    public void objectSummaryLimitTests() {
        long now = System.currentTimeMillis();
        List<S3ObjectSummary> summaries = Arrays.asList(
                summary(3, "a"),
                summary(3, "b"),
                summary(3, "c"),
                summary(2, "d"),
                summary(2, "e"),
                summary(2, "f"),
                summary(1, "g"),
                summary(1, "h"),
                summary(1, "j")
        );

        assertEquals(3, SubmissionFromS3Repository.limit(summaries, 1, 5).size());
        assertEquals(3, SubmissionFromS3Repository.limit(summaries, 2, 5).size());
        assertEquals(3, SubmissionFromS3Repository.limit(summaries, 3, 5).size());
        assertEquals(5, SubmissionFromS3Repository.limit(summaries, 4, 5).size());
        assertEquals(5, SubmissionFromS3Repository.limit(summaries, 5, 5).size());
        assertEquals(5, SubmissionFromS3Repository.limit(summaries, 6, 5).size());

        assertEquals(9, SubmissionFromS3Repository.limit(summaries, 7, 100).size());
    }

    private S3ObjectSummary summary(int ageSeconds, String key) {
        S3ObjectSummary summary = new S3ObjectSummary();
        summary.setLastModified(new Date(now - ageSeconds * 1000));
        summary.setKey(key);
        return summary;
    }
}

File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/diagnosiskeydist/SubmissionRepositoryTest.kt
package uk.nhs.nhsx.diagnosiskeydist

import com.amazonaws.services.s3.model.S3ObjectSummary
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import uk.nhs.nhsx.diagnosiskeydist.s3.SubmissionFromS3Repository
import uk.nhs.nhsx.diagnosiskeyssubmission.model.StoredTemporaryExposureKey
import uk.nhs.nhsx.diagnosiskeyssubmission.model.StoredTemporaryExposureKeyPayload
import uk.nhs.nhsx.testhelper.data.TestData
import java.io.ByteArrayInputStream
import java.time.Instant
import java.util.*

class SubmissionRepositoryTest {

    private val now = Instant.parse("2020-01-13T13:00:00.000Z")

    @Test
    fun deserializesStoredPayload() {
        val inputStream = ByteArrayInputStream(TestData.STORED_KEYS_PAYLOAD.toByteArray())
        val payload = SubmissionRepository.getTemporaryExposureKeys(inputStream)
        assertThat(payload.temporaryExposureKeys).isNotEmpty
        assertThat(payload.temporaryExposureKeys.stream().allMatch { matchKey(it, TestData.STORED_KEYS_PAYLOAD_DESERIALIZED) }).isTrue
    }

    @Test
    fun deserializesStoredPayloadWithDaysSinceOnset() {
        val inputStream = ByteArrayInputStream(TestData.STORED_KEYS_PAYLOAD_DAYS_SINCE_ONSET.toByteArray())
        val payload = SubmissionRepository.getTemporaryExposureKeys(inputStream)
        assertThat(payload.temporaryExposureKeys).isNotEmpty
        assertThat(payload.temporaryExposureKeys.stream().allMatch { matchKey(it, TestData.STORED_KEYS_PAYLOAD_DESERIALIZED_DAYS_SINCE_ONSET) }).isTrue
    }

    @Test
    fun `limits submissions`() {
        val summaries = listOf(
            summary(2, "e", now),
            summary(2, "f", now),
            summary(2, "d", now),
            summary(3, "a", now),
            summary(3, "b", now),
            summary(3, "c", now),
            summary(1, "h", now),
            summary(1, "g", now),
            summary(1, "j", now),
        )

        val testcases = listOf(
            Triple(1, 5, listOf("a", "b", "c")),
            Triple(2, 5, listOf("a", "b", "c")),
            Triple(3, 5, listOf("a", "b", "c")),
            Triple(4, 5, listOf("a", "b", "c", "e", "f")),
            Triple(5, 5, listOf("a", "b", "c", "e", "f")),
            Triple(6, 5, listOf("a", "b", "c", "e", "f")),
            Triple(7, 100, listOf("a", "b", "c", "e", "f", "d", "h", "g", "j")),
        )

        testcases.forEach { (limit, maxResults, expected) ->
            assertThat(SubmissionFromS3Repository.limit(summaries, limit, maxResults))
                .describedAs("limit: $limit, maxResults: $maxResults")
                .extracting("key")
                .containsExactlyElementsOf(expected)
        }
    }

    @Test
    fun `throws exception if limit is 0`() {
        val summaries = listOf(summary(2, "e", now))

        assertThatThrownBy { SubmissionFromS3Repository.limit(summaries, 0, 10) }
            .hasMessage("limit needs to be greater than 0")
    }

    @Test
    fun `throws exception if limit is negative`() {
        val summaries = listOf(summary(2, "e", now))

        assertThatThrownBy { SubmissionFromS3Repository.limit(summaries, -1, 10) }
            .hasMessage("limit needs to be greater than 0")
    }

    private fun summary(ageSeconds: Long, objectKey: String, now: Instant): S3ObjectSummary =
        S3ObjectSummary().apply {
            key = objectKey
            lastModified = Date.from(now.minusSeconds(ageSeconds))
        }

    private fun matchKey(storedKey: StoredTemporaryExposureKey, deserializedPayload: StoredTemporaryExposureKeyPayload): Boolean =
        deserializedPayload.temporaryExposureKeys.stream()
            .anyMatch { k: StoredTemporaryExposureKey ->
                k.key == storedKey.key
                    && k.rollingPeriod == storedKey.rollingPeriod
                    && k.rollingStartNumber == storedKey.rollingStartNumber
                    && k.transmissionRisk == storedKey.transmissionRisk
                    && (k.daysSinceOnsetOfSymptoms == null && storedKey.daysSinceOnsetOfSymptoms == null || k.daysSinceOnsetOfSymptoms != null && k.daysSinceOnsetOfSymptoms == storedKey.daysSinceOnsetOfSymptoms)
            }
}
